### ISSUE TITLE
Unable to install OTP 20.2+ On CentOS 6

### ISSUE THREAD

**Issue by nalundgaard** [kerl/kerl#279]
With kerl 1.8.2 on a CentOS 6.9 box (one of our Jenkins slaves), I am getting a strange error when building OTP 20.2.x/20.3.x:

```
$ /home/jenkins/bin/kerl build 20.2.4 20.2.4
Extracting source code
Building Erlang/OTP 20.2.4 (20.2.4), please wait...
which: no dpkg in (/home/jenkins/erlang/19.3.6.8/bin:/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin:/root/bin:/home/jenkins/bin)
APPLICATIONS DISABLED (See: /home/jenkins/.kerl/builds/20.2.4/otp_build_20.2.4.log)
 * jinterface     : No Java compiler found
 * odbc           : odbc disabled by user.
 * odbc           : User gave --without-odbc option

APPLICATIONS INFORMATION (See: /home/jenkins/.kerl/builds/20.2.4/otp_build_20.2.4.log)
 * wx             : wxWidgets not found, wx will NOT be usable

DOCUMENTATION INFORMATION (See: /home/jenkins/.kerl/builds/20.2.4/otp_build_20.2.4.log)
 * documentation  :
 *                  xsltproc is missing.
 *                  fop is missing.
 *                  The documentation can not be built.

Build failed.
/usr/include/openssl/ec.h:82:2: error: #error EC is disabled.
gmake[4]: *** [../priv/obj/x86_64-unknown-linux-gnu/crypto.o] Error 1
gmake[4]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
gmake[3]: *** [opt] Error 2
gmake[3]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
gmake[2]: *** [opt] Error 2
gmake[2]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto'
gmake[1]: *** [opt] Error 2
gmake[1]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib'
gmake: *** [libs] Error 2

Please see /home/jenkins/.kerl/builds/20.2.4/otp_build_20.2.4.log for full details.
```

This is the contents of `~/.kerlrc`:

```sh
export CFLAGS="-DOPENSSL_NO_EC=1"
KERL_CONFIGURE_OPTIONS="--enable-smp-support --enable-threads --enable-kernel-poll  --enable-hipe --enable-m64-build"
KERL_CONFIGURE_DISABLE_APPLICATIONS=odbc
KERL_BUILD_BACKEND=git
OTP_GITHUB_URL="https://github.com/erlang/otp"
```

Here's a fuller output from the build log:

```
$ tail -n 20 /home/jenkins/.kerl/builds/20.2.4/otp_build_20.2.4.log
 VSN	../ebin/crypto.app
 VSN	../ebin/crypto.appup
gmake[3]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/src'
gmake[3]: Entering directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
 MAKE	opt
gmake[4]: Entering directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
 CC	../priv/obj/x86_64-unknown-linux-gnu/crypto.o
In file included from /usr/include/openssl/ecdh.h:78,
                 from /usr/include/openssl/engine.h:85,
                 from crypto.c:63:
/usr/include/openssl/ec.h:82:2: error: #error EC is disabled.
gmake[4]: *** [../priv/obj/x86_64-unknown-linux-gnu/crypto.o] Error 1
gmake[4]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
gmake[3]: *** [opt] Error 2
gmake[3]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto/c_src'
gmake[2]: *** [opt] Error 2
gmake[2]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib/crypto'
gmake[1]: *** [opt] Error 2
gmake[1]: Leaving directory `/home/jenkins/.kerl/builds/20.2.4/otp_src_20.2.4/lib'
gmake: *** [libs] Error 2
```

It's almost like $CFLAGS isn't getting propagated down for the compilation, but the logs clearly indicate that it is. Would that be an Erlang/OTP bug/regression?

I've attached the build log for anyone curious. Note that this particular log is for an attempt at getting 20.3.5 running, with the same result as 20.2.4.

[otp_build_20.3.5.log](https://github.com/kerl/kerl/files/1973328/otp_build_20.3.5.log)

**Comment by nalundgaard** [comment]
@danilagamma kindly pointed me to this Erlang-questions thread that also showed up in my daily digest just today. Hopefully the fix comes soon.

[[centos-7]With newer Erlang versions has the eliptic curve crypto situation changed?](http://erlang.org/pipermail/erlang-questions/2018-May/095368.html)

There's nothing for kerl to do here, except maybe note this in the README under the section on Centos & EC, but I think we should leave this open till this is patched in OTP 20.x/21-rc2.

**Comment by nalundgaard** [comment]
So, I think maybe this is a tempest in a teacup? It seems like with Elang 20.2+, you don't need `CFLAGS="-DOPENSSL_NO_EC=1"` to get a working OTP installation? I commented out that line of my `.kerlrc` on our CentOS 6.9 box and was able to build and install Erlang 20.2.4.

Here's the box's openssl data:

```
[jenkins@localhost ~]$ openssl ecparam -list_curves
  secp384r1 : NIST/SECG curve over a 384 bit prime field
  secp521r1 : NIST/SECG curve over a 521 bit prime field
  prime256v1: X9.62/SECG curve over a 256 bit prime field
[jenkins@localhost ~]$ openssl version
OpenSSL 1.0.1e-fips 11 Feb 2013
[jenkins@localhost ~]$ yum list installed | grep openssl
openssl.x86_64                       1.0.1e-57.el6                    @base
openssl-devel.x86_64                 1.0.1e-57.el6                    @base
```

I wrote this little tester for EC:

```erlang
VerifyCurve = fun(Curve) ->
    Msg = <<"hello world!">>,
    try
        {PublicKey, PrivKeyOut} = crypto:generate_key(ecdh,Curve),
        Signature = crypto:sign(ecdsa, sha512, Msg, [PrivKeyOut, Curve]),
        true = crypto:verify(ecdsa, sha512, <<"hello world!">>, Signature, [PublicKey, Curve]),
        {ok, Curve}
    catch
       Err:Reason ->
           {error, {Curve, {Err, Reason}}}
    end
end.

rp([VerifyCurve(C) || C <- crypto:ec_curves()]).
```

Result on the box:

```erlang
2> rp([VerifyCurve(C) || C <- crypto:ec_curves()]).
[{error,{secp112r1,{error,badarg}}},
 {error,{secp112r2,{error,badarg}}},
 {error,{secp128r1,{error,badarg}}},
 {error,{secp128r2,{error,badarg}}},
 {error,{wtls6,{error,badarg}}},
 {error,{wtls8,{error,badarg}}},
 {error,{secp160k1,{error,badarg}}},
 {error,{secp160r1,{error,badarg}}},
 {error,{secp160r2,{error,badarg}}},
 {error,{secp192r1,{error,badarg}}},
 {error,{secp192k1,{error,badarg}}},
 {error,{secp224k1,{error,badarg}}},
 {error,{secp224r1,{error,badarg}}},
 {ok,secp256k1},
 {ok,secp256r1},
 {ok,secp384r1},
 {ok,secp521r1},
 {error,{prime192v1,{error,badarg}}},
 {error,{prime192v2,{error,badarg}}},
 {error,{prime192v3,{error,badarg}}},
 {error,{prime239v1,{error,badarg}}},
 {error,{prime239v2,{error,badarg}}},
 {error,{prime239v3,{error,badarg}}},
 {ok,prime256v1},
 {error,{wtls7,{error,badarg}}},
 {error,{wtls9,{error,badarg}}},
 {error,{wtls12,{error,badarg}}},
 {error,{brainpoolP160r1,{error,badarg}}},
 {error,{brainpoolP160t1,{error,badarg}}},
 {error,{brainpoolP192r1,{error,badarg}}},
 {error,{brainpoolP192t1,{error,badarg}}},
 {error,{brainpoolP224r1,{error,badarg}}},
 {error,{brainpoolP224t1,{error,badarg}}},
 {ok,brainpoolP256r1},
 {ok,brainpoolP256t1},
 {ok,brainpoolP320r1},
 {ok,brainpoolP320t1},
 {ok,brainpoolP384r1},
 {ok,brainpoolP384t1},
 {ok,brainpoolP512r1},
 {ok,brainpoolP512t1}]
ok
```

Seems that CentOS's openSSL isn't lying.

**Comment by mrallen1** [comment]
Good to know. Thanks.

**Comment by nalundgaard** [comment]
As a follow-up to this, have a look at [ERL-823](https://bugs.erlang.org/browse/ERL-823). It seems that as of OTP 17+, we no longer need to use `-DOPENSSL_NO_EC=1` (can use `-DOPENSSL_NO_EC2M=1`, but only needed of building an OTP installation on a non RHEL/CentOS box to be deployed on one).

This does result in a functional ssl/crypto module, with the caveat that, as I observed above, the `ssl:eccs/0` and `crypto:ec_curves/0` both return unsupported curves. Since the first 7 curves returned by `ssl:eccs/0` *are* supported, the `ssl` module should work for the most part, with the exception of negotiation with peers who don't support those first 7 curves. It's probably worth updating the documentation accordingly.

**Comment by fenollp** [comment]
I wonder if kerl should be at least pointing to this issue when the tuple otp version / distro / exported env var matches?
I’d love this sort of thing really! But surely this wouldn’t work when building shas instead of tags. Would you guys be fine with that?
@mrallen1 @sanmiguel

**Comment by fenollp** [comment]
@nalundgaard Could you make a PR against the README plz?

**Comment by mrallen1** [comment]
@fenollp I asked Nicholas to add a comment here for tracking purposes, so I don't expect a PR from him but rather from me :)  As for your comment above - I'm not sure what you mean. That if kerl is building on a Red Hat-ish system and it's building OTP >17 and "OPENSSL_NO_EC" is defined, it should emit a warning?

**Comment by nalundgaard** [comment]
Looks like maybe a README update is premature. It looks like erlang/otp#2085 should fully resolve this issue, and once it's resolved, we can close this issue (and #212) permanently, though it may require backporting the patch to older OTP releases (or just leaving a warning in the README about them).
